pybamm_tea/tea.py

```diff
@@ -394,223 +425,191 @@ def print_capacities_and_potentials(self):
-                "Unit": "-",
-                "Value": stack_ed.get("Positive electrode stoichiometry at 0% SoC"),
-            },
+        stack_ed = self.stack_energy
+
+        parameters = [
+            "Stack average OCP [V]",
+            "Minimal OCP [V]",
+            "Maximal OCP [V]",
+            "Negative electrode average OCP [V]",
+            "Positive electrode average OCP [V]",
+            "Practical n/p ratio",
+            "Theoretical n/p ratio",
+            "Volumetric stack capacity [A.h.L-1]",
+            "Gravimetric stack capacity [A.h.kg-1]",
+            "Capacity [mA.h.cm-2]",
+            "Lithium inventory [mA.h.cm-2]",
+            "Initial loss of lithium inventory [mA.h.cm-2]",
+            "Initial lithium inventory [mA.h.cm-2]",
+            "Initial lithium inventory / positive electrode theoretical capacity",
+            "Negative electrode theoretical capacity [mA.h.cm-2]",
+            "Positive electrode theoretical capacity [mA.h.cm-2]",
+            "Negative electrode volumetric capacity [A.h.L-1]",
+            "Positive electrode volumetric capacity [A.h.L-1]",
+            "Negative electrode gravimetric capacity [A.h.kg-1]",
+            "Positive electrode gravimetric capacity [A.h.kg-1]",
+            "Negative electrode active material practical capacity [A.h.kg-1]",
+            "Positive electrode active material practical capacity [A.h.kg-1]",
+            "Negative electrode active material capacity [A.h.kg-1]",
+            "Positive electrode active material capacity [A.h.kg-1]",
+            "Negative electrode stoichiometry at 0% SoC",
+            "Negative electrode stoichiometry at 100% SoC",
+            "Positive electrode stoichiometry at 100% SoC",
+            "Positive electrode stoichiometry at 0% SoC",
         ]
-        # Create the DataFrame from the list of dictionaries
+
+        names = []
+        units = []
+        values = []
+        for parameter in parameters:
+            if "[" in parameter:
+                name, unit = parameter.split("[")
+                names.append(name.strip())
+                units.append(unit.rstrip(']'))
+            else:
+                names.append(parameter.strip())
+                units.append("-")
+            values.append(stack_ed.get(parameter))
+
+        data = {
+            "Parameter": names,
+            "Unit": units,
+            "Value": values,
+            }
+        
         df = pd.DataFrame(data)
         return df
-
-    def calculate_stack_energy_densities(self):
+        
+    
+    def calculate_stack_energy(self):
         """
         Calculate ideal volumetric and gravimetric energy densities on stack level.
         """
         stack_ed = {}  # stack energy densities dict
-        pava = None
-        pava = self.parameter_values  # parameter values
+        pava = dict(self.parameter_values)  # parameter values
 
         # stoichiometries - calculation based on input stoichiometries or cell
```

Comment:
there is a lot of repeated code here and it is hard to follow. can you explain the use case for being able to independently change sto limits?

Reply:
It is useful for capacity balancing in case no OCV curves are supplied. I also generally like it to calculate voltage cut-offs based on it. I'll try to shorten the code section and add more comments so that it becomes easier to follow.